### DISCUSSION TITLE
Server has closed the connection in long transactions with ?pgbouncer=true

### DISCUSSION THREAD

**Discussion by AurelienGasser** [prisma/prisma#23825] [Q&A]
### Bug description

When using `?pgbouncer=true`, long transactions fail after about 20 seconds with error

```
Invalid `prisma.$executeRaw()` invocation:

Server has closed the connection.
```

This happens even with a long timeout setting for `prisma.$transaction`.

- Removing `?pgbouncer=true` fixes the issue.
- Ssetting `?connection_limit=1` also fixes the issue

Additional notes:
- I cannot repro from my local machine, only on Vercel
- I cannot repro connecting using a direct SQL connection, only connecting to a pooler triggers triggers the bug
- There are no other clients (isolated DB / Vercel deployment)

```
Env     | Use transaction | Connection                               | Result
----------------------------------------------------------------------------------
laptop    NO                Direct                                     no bug
laptop    NO                Pooler (Supavisor) ?pgbouncer=true         no bug
laptop    YES               Direct                                     no bug
laptop    YES               Pooler (Supavisor)                         no bug

Vercel    NO                Direct                                     no bug
Vercel    YES               Direct                                     no bug
Vercel    YES               Direct                                     no bug
Vercel    YES               Direct             ?connection_limit=1     no bug
Vercel    YES               Pooler (Supavisor) ?connection_limit=1     no bug
Vercel    YES               Pooler (Supavisor) ?pgbouncer=true         timeout after 20 sec
```

### How to reproduce

```ts
const start = new Date();

try {
  const prisma = new PrismaClient({ datasourceUrl: url });
  await prisma.$transaction(
    async (tx) => {
      for (let i = 0; i < 100000; ++i) {
        await tx.$executeRaw`INSERT INTO foo (created_on,name) VALUES (${new Date()},${`foo number ${i}`})`;
        console.log(i);
      }
      await tx.$executeRaw`ROLLBACK`;
    },
    { timeout: 60_000 }
  );
} finally {
  console.log('operation took', (new Date().getTime() - start.getTime()) / 1000, 's');
}
```

### Expected behavior

The transaction should either succeed or timeout.

### Prisma information

```prisma
datasource db {
  url       = env("DATABASE_URL")
  provider  = "postgresql"
}

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["rhel-openssl-1.0.x"]
}

model Pen {
  id        Int          @id @default(autoincrement())
  createdOn DateTime     @default(now())
  name      String
}
```

### Environment & setup

- Platform: Vercel
- Database: PostgreSQL (Supabase)

### Prisma Version

```
5.12.1
```

**Comment by ludralph** [comment]
Hi @AurelienGasser 👋 

Thank you for raising this question.

The recommended set up for using pgbouncer from version 1.21.0 is shown below
```
max_prepared_statements>0
pool_mode=transaction
remove pgbouncer=true
```

You can see the issue comment for more information and context

- https://github.com/prisma/prisma/issues/21531#issuecomment-1891628066

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
